File: solver.go
// -*- coding:utf-8 -*-

//
// Pentomino Puzzle Solver with Go
//

package main

import (
  "os"
  "fmt"
  "sort"
  "strings"
  "strconv"
)

const PIECE_DEF_DOC =`
+-------+-------+-------+-------+-------+-------+
|       |   I   |  L    |  N    |       |       |
|   F F |   I   |  L    |  N    |  P P  | T T T |
| F F   |   I   |  L    |  N N  |  P P  |   T   |
|   F   |   I   |  L L  |    N  |  P    |   T   |
|       |   I   |       |       |       |       |
+-------+-------+-------+-------+-------+-------+
|       | V     | W     |   X   |    Y  | Z Z   |
| U   U | V     | W W   | X X X |  Y Y  |   Z   |
| U U U | V V V |   W W |   X   |    Y  |   Z Z |
|       |       |       |       |    Y  |       |
+-------+-------+-------+-------+-------+-------+
`

type Point struct { x int;  y int }
type Fig []Point

const SPACE = ' '

var ELEMS [][]string
var debug_flg bool

func piece_def( id rune) []Point {
  var ret [] Point
  x := 0;   y := 0
  for _, ch := range PIECE_DEF_DOC {
    if  ch == id  { ret = append( ret, Point{ x/2, y } )  }

    if ch == '\n' { y++; x = 0
    } else        { x++ }
  }
  return ret
}

func init() {
  //         2
  // (-1,-1) | (0,-1)
  //   ---4--+--1----
  // (-1, 0) | (0, 0)
  //         8
  elem := []string {
    //0      3     5    6    7     9    10   11   12   13   14   15
    "    ,,,+---,,----,+   ,+---,,+---,|   ,+---,+   ,+---,+   ,+---",
    "    ,,,    ,,    ,    ,    ,,|   ,|   ,|   ,|   ,|   ,|   ,|   ",
  }

  ELEMS = [][]string{
    strings.Split( elem[0], "," ),
    strings.Split( elem[1], "," ),
  }

  debug_flg = false
}

///////////////////////////////////////////////////////////////////

type Piece struct {
  next *Piece
  id   rune
  figs []Fig
}


func NewPiece( id rune, fig_def Fig, next *Piece ) *Piece{

  self := Piece{ id: id, next: next }

  for r_f := 0; r_f < 8; r_f++ {
    var fig []Point
    for _, pt := range fig_def {
      for t := 0; t < ( r_f % 4 ); t++ { pt.x, pt.y = -pt.y, pt.x }
      if r_f >= 4                      { pt.x       = -pt.x       }
      fig = append( fig, pt )
    }

    //sort
    sort.Slice( fig, func( i, j int ) bool {
      if fig[i].y == fig[j].y { return fig[i].x < fig[j].x }
      return fig[i].y < fig[j].y
    })

    // normalize
    for i := len( fig ) - 1; i >= 0; i--  {
      fig[i].x -= fig[0].x
      fig[i].y -= fig[0].y
    }

    // uniq
    if ! strings.Contains( fmt.Sprint(self.figs), fmt.Sprint(fig) ) {
      self.figs = append( self.figs, fig )
    }
  }

  if debug_flg {
    fmt.Printf( "%s: (%d)\n", string(id), len( self.figs ) )
    for _, fig := range self.figs {
      fmt.Printf( "    %s\n", fmt.Sprint(fig) )
    }
  }
  return &self
}

///////////////////////////////////////////////////////////////////

type Board struct {
  cells  [][]rune
  width  int
  height int
}


func NewBoard( width int,  height int ) *Board {
  self := Board{ width: width, height: height }

  for i := 0; i < self.height; i++  {
    var line  []rune
    for j := 0; j < self.width; j++ {
      line = append( line, SPACE )
    }
    self.cells= append( self.cells, line )
  }
  if width * height == 64 {      // 8x8 or 4x16
    self.cells[ height/2 - 1 ][ width/2 - 1] = '@'
    self.cells[ height/2 - 1 ][ width/2    ] = '@'
    self.cells[ height/2     ][ width/2 - 1] = '@'
    self.cells[ height/2     ][ width/2    ] = '@'
  }

  return &self
}


func (self *Board)render() string {
  var lines []string
  for y := 0; y <= self.height; y++ {
    for d := 0; d < 2; d++ {
      line := ""
      for x := 0; x <= self.width; x++ {
        c := 0
        if self.at( x+0, y+0 ) != self.at( x+0, y-1 ) { c += 1 }
        if self.at( x+0, y-1 ) != self.at( x-1, y-1 ) { c += 2 }
        if self.at( x-1, y-1 ) != self.at( x-1, y+0 ) { c += 4 }
        if self.at( x-1, y+0 ) != self.at( x+0, y+0 ) { c += 8 }
        line += ELEMS[ d ][ c ]
      }
      lines = append( lines, line )
    }
  }
  return strings.Join( lines, "\n" )

}


func (self *Board)at( x int, y int ) rune {
  if x >= 0 && x < self.width && y >= 0 && y < self.height {
    return self.cells[y][x]
  }
  return '?'
}


func (self *Board)check( ox int, oy int, fig Fig ) bool {
  for i := range fig  {
    if self.at( ox + fig[i].x, oy + fig[i].y ) != SPACE {
      return false
    }
  }
  return true
}


func (self *Board)place( ox int, oy int, fig Fig, id rune ) {
  for i := range fig  {
    self.cells[ oy + fig[i].y ][ ox + fig[i].x ] = id
  }
}


func(self *Board)find_space( x int, y int ) Point {
  for  self.cells[y][x] != SPACE  {
    x = ( x + 1 ) % self.width
    if x == 0  { y++ }
  }
  return Point{ x, y }
}

///////////////////////////////////////////////////////////////////

type Solver struct {
    Unused  *Piece
    board *Board
    solutions int
}


func NewSolver( width int,  height int ) *Solver {
  self := Solver{ solutions: 0 }

  var pc *Piece = nil
  ids := []rune( "FLINPTUVWXYZ" )         // the first 'X' is dummy
  for i := len(ids) - 1; i >= 0; i--  {
    pc = NewPiece( ids[i], piece_def( ids[i] ), pc )
  }
  self.Unused = NewPiece( '!', [] Point{}, pc )
  // limit the symmetry of 'F'
  if width == height {
    pc.figs = pc.figs[0:1]
  } else {
    pc.figs = pc.figs[0:2]
  }

  self.board = NewBoard( width, height )

  return &self
}


func (self *Solver)solve( x int, y int ) {
  if  self.Unused.next != nil {
    xy   := self.board.find_space( x, y )
    x = xy.x
    y = xy.y                              // update x,y
    prev := self.Unused
    for prev.next != nil {
      pc := prev.next
      prev.next = pc.next
      for _, fig := range pc.figs  {
        if self.board.check( x, y, fig ) == true {
          self.board.place( x, y, fig, pc.id )
          self.solve( x, y )
          self.board.place( x, y, fig, SPACE )
        }
      }
      prev.next = pc
      prev      = pc
    }
  } else {
    self.solutions++
    curs_up := ""
    if self.solutions > 1 { curs_up = fmt.Sprintf( "\033[%dA", self.board.height * 2 + 2 ) }
    fmt.Printf( "%s%s%d\n", curs_up, self.board.render(), self.solutions )
  }
}

///////////////////////////////////////////////////////////////////


func main() {
  width  := 6
  height := 10
  for _, arg := range os.Args {
    if arg == "--debug" { debug_flg = true }
    ar := strings.Split( arg, "x" )
    if len(ar) == 2 {
      w, err := strconv.Atoi( ar[0] )
      if err == nil {
        h, err := strconv.Atoi( ar[1] )
        if err == nil && w >=3 && h >= 3 && ( w * h == 60 || w * h == 64 ) {
          width  = w
          height = h
        }
      }
    }
  }

  solver := NewSolver( width, height )
  solver.solve( 0, 0 )
}

File: solver.nim
# -*- coding:utf-8 -*-

#
# Pentomino Puzzle Solver with Nim
#
import os, sequtils, re, strutils, algorithm, strformat
import tables, sugar

const PIECE_DEF_DOC = """
+-------+-------+-------+-------+-------+-------+
|       |   I   |  L    |  N    |       |       |
|   F F |   I   |  L    |  N    |  P P  | T T T |
| F F   |   I   |  L    |  N N  |  P P  |   T   |
|   F   |   I   |  L L  |    N  |  P    |   T   |
|       |   I   |       |       |       |       |
+-------+-------+-------+-------+-------+-------+
|       | V     | W     |   X   |    Y  | Z Z   |
| U   U | V     | W W   | X X X |  Y Y  |   Z   |
| U U U | V V V |   W W |   X   |    Y  |   Z Z |
|       |       |       |       |    Y  |       |
+-------+-------+-------+-------+-------+-------+
"""

type
  Point = tuple[ x: int, y: int ]
  Fig   = seq[ Point ]

var PIECE_DEF = initTable[ char, Fig ]()

for y, line in PIECE_DEF_DOC.splitLines().pairs():
  for x, c in line.pairs():
    if $c =~ re"\w":
      PIECE_DEF.mgetOrPut( c, @[] ).add( ( x div 2, y ) )

var debug_flg = false

#############################################################
#
# Piece
#
type
  Piece = ref object
    id:   char
    figs: seq[ Fig ]
    next: Piece


proc newPiece( id: char, pc_def: Fig, next: Piece ): Piece =
  var figs: seq[ Fig ] = @[]
  for rf in 0 ..< 8:
    var fig = pc_def;
    for _ in 0 ..< rf mod 4: fig = fig.mapIt( ( -it.y, it.x ) ) # rotate
    if rf >= 4:              fig = fig.mapIt( ( -it.x, it.y ) ) # flip
    fig.sort( (a, b:Point) =>  cmp((a.y, a.x), (b.y, b.x)) )    # sort
    fig = fig.mapIt( ( it.x - fig[0].x, it.y - fig[0].y ) )     # normalize
    if fig notin figs: figs.add( fig )                          # uniq

  if debug_flg:
    var n = figs.len()
    echo fmt"{id}: ({n})"
    for pts in figs:
      var str = pts.mapIt( fmt"({it.x},{it.y})" ).join(", " )
      echo "\t", fmt"[ {str} ]"

  Piece( id: id, figs: figs, next: next )

#############################################################
#
# Board
#
type
  Board = ref object
    width:  int
    height: int
    cells:  seq[seq[char]]

let SPACE = ' '
method place( o: Board, x: int, y: int, fig: Fig, id: char ) {.base.}

proc newBoard( w, h: int ): Board =
  let cells = newSeqWith( h, newSeqWith( w, SPACE ) )
  result = Board( width: w, height: h, cells: cells )
  if w * h == 64:            # 8x8 or 4x16
    result.place( ( w div 2 ) - 1, ( h div 2 ) - 1,
                  @[ (0,0), (0,1), (1,0), (1,1) ], '@' )


method at( o:Board, x: int, y: int ): char {.base.} =
  if x >= 0 and x < o.width and y >= 0 and y < o.height:
    o.cells[y][x]
  else:
    '?'


method check( o: Board, x: int, y: int, fig: Fig ): bool {.base.} =
  fig.allIt( o.at( x + it.x, y + it.y ) == SPACE )


method place( o: Board, x: int, y: int, fig: Fig, id: char ) {.base.} =
  for pt in fig:
    o.cells[ y + pt.y ][ x + pt.x ] = id


method find_space( o: Board, xx: int, yy: int ): Point {.base.} =
  var ( x, y ) = ( xx, yy )
  while o.cells[ y ][ x ] != SPACE:
    inc( x )
    if x == o.width:
      x = 0
      inc( y )
  ( x, y )


let ELEMS = @[
  "    ,,,+---,,----,+   ,+---,,+---,|   ,+---,+   ,+---,+   ,+---",
  "    ,,,    ,,    ,    ,    ,,|   ,|   ,|   ,|   ,|   ,|   ,|   "
].mapIt( it.split( ',' ) )

method render( o: Board ): string {.base.} =
  toSeq( 0 .. o.height ).map( proc (y: int) : string =
    var codes = toSeq( 0..o.width ).map( proc (x: int): int =
      var ids = @[(0,0),(0,1),(1,1),(1,0)].mapIt( o.at( x-it[0], y-it[1] ) )
      ( ( if ids[0] != ids[1]: 1 else: 0 ) +
        ( if ids[1] != ids[2]: 2 else: 0 ) +
        ( if ids[2] != ids[3]: 4 else: 0 ) +
        ( if ids[3] != ids[0]: 8 else: 0 ) ) )
    ELEMS.map( (elem: seq[string]) => codes.map( (c: int) => elem[c] ).join()
    ).join( "\n" )
  ).join( "\n" )


#############################################################
#
# Solver
#
type
  Solver = ref object
    board:     Board
    unused:    Piece
    solutions: int


proc newSolver( width:int, height:int ): Solver =
  let board = newBoard( width, height )

  var pc: Piece = nil
  for id in "FLINPTUVWXYZ".toSeq.reversed():
    pc = newPiece( id, PIECE_DEF[ id ], pc )

  var unused = newPiece( '!', @[], pc )    # dummy piece
  # limit the symmetry of 'F'
  pc.figs = pc.figs[ 0.. (if width == height: 0 else: 1) ]

  Solver( board: board, solutions: 0, unused: unused )


method solve( o: Solver, x: int, y: int ) {.base.} =
  if o.unused.next != nil:
    let ( x, y ) = o.board.find_space( x, y )
    var prev = o.unused
    while prev.next != nil:
      var pc = prev.next
      prev.next = pc.next
      for fig in pc.figs:
        if o.board.check( x, y, fig ):
          o.board.place( x, y, fig, pc.id )
          o.solve( x, y )                 # call recursively
          o.board.place( x, y, fig, SPACE )
      prev.next = pc
      prev      = pc
  else:
    inc( o.solutions )
    let cursup = if o.solutions > 1:
                   fmt"{'\x1b'}[{ o.board.height * 2 + 2 }A"
                 else:
                   ""
    echo cursup, o.board.render(), o.solutions

#############################################################

var (width, height) = (6, 10)

for arg in commandLineParams():
  if arg == "--debug":  debug_flg = true

  if arg =~ re"(\d+)\D(\d+)":
     let ( w, h ) = ( parseInt( matches[0] ), parseInt( matches[1] ) )
     if w >= 3 and h >= 3 and ( w * h == 60 or w * h == 64 ):
       ( width, height ) = ( w, h )

let solver = newSolver( width, height )
solver.solve( 0, 0 )
